src/Microsoft.OData.Core/UriParser/SemanticAst/FilterSegment.cs

```diff
@@ -0,0 +1,150 @@
+    public sealed class FilterSegment : ODataPathSegment
+    {
+        /// <summary>
+        /// The filter expression - this should evaluate to a single boolean value.
+        /// </summary>
+        private readonly SingleValueNode expression;
+
+        /// <summary>
+        /// The parameter for the alias which represents a single value from the collection.
+        /// </summary>
+        private readonly RangeVariable rangeVariable;
+
+        /// <summary>
+        /// The type that this segment is bound to.
+        /// </summary>
+        private readonly IEdmType bindingType;
+
+        /// <summary>
+        /// String representing "$filter(expression)".
+        /// </summary>
+        private readonly string fullSegment;
+
+        /// <summary>
+        /// Build a segment representing $filter.
+        /// </summary>
+        /// <param name="expression">The filter expression - this should evaluate to a single boolean value.</param>
+        /// <param name="rangeVariable">An expression that represents a single value from the collection.</param>
+        /// <param name="navigationSource">The navigation source that this filter applies to.</param>
+        /// <exception cref="System.ArgumentNullException">Throws if any input parameter is null.</exception>
+        /// <remarks>$filter should not be applied on singletons or single entities.</remarks>
+        public FilterSegment(SingleValueNode expression, RangeVariable rangeVariable, IEdmNavigationSource navigationSource)
+        {
+            ExceptionUtils.CheckArgumentNotNull(expression, "expression");
+            ExceptionUtils.CheckArgumentNotNull(rangeVariable, "rangeVariable");
+            ExceptionUtils.CheckArgumentNotNull(navigationSource, "navigationSource");
+
+            this.Identifier = UriQueryConstants.FilterSegment;
+            this.SingleResult = false;
+            this.TargetEdmNavigationSource = navigationSource;
+            this.TargetKind = RequestTargetKind.Resource;
+            this.TargetEdmType = rangeVariable.TypeReference.Definition;
+
+            this.expression = expression;
+            this.rangeVariable = rangeVariable;
+            this.bindingType = navigationSource.Type;
+
+            NodeToStringBuilder nodeToStringBuilder = new NodeToStringBuilder();
+            string expressionString = nodeToStringBuilder.TranslateNode(expression);
+            this.fullSegment = UriQueryConstants.FilterSegment + ExpressionConstants.SymbolOpenParen + expressionString
+                + ExpressionConstants.SymbolClosedParen;
+        }
+
+        /// <summary>
+        /// Gets the filter expression - this should evaluate to a single boolean value.
+        /// </summary>
+        public SingleValueNode Expression
+        {
+            get { return this.expression; }
+        }
+
+        /// <summary>
+        /// Gets the string representing "$filter(expression)".
+        /// </summary>
+        public string FullSegment

```

Comment:
>FullSegment [](start = 22, length = 11)

Should this be "LiteralText" (similar to PathTemplateSegment)?  Should it be public (versus used for internally, i.e. for serialization to a string)?

Reply:
Good idea. Changing to LiteralText and making it internal.

Reply:
Update: upon making it internal, a test service `test\FunctionalTests\Service\Microsoft.OData.Service.csproj` has caused a lot of issues. (Not sure if I really like the way that the test service is designed.)

1. The test service tried to access some of the internals of the source code (specifically the new internal LiteralText).
2. Upon either enabling internals to the test service or adding a link of FilterSegment.cs to the test service, the test service tries to access more internals.
3. Upon adding those internals that the test service tries to access, the chain of access to internals  continues to grow, so I abandoned this approach because it looked like it was going to take a while to add everything needed.
---
1. I also tried to move the LiteralText building into the Translator classes. However, due to the need of using the NodeToStringBuilder (which is also internal) to help us determine the text, we come across the same problem as the one mentioned above. The Microsoft.OData.Services project will not build due to the lack of access to the internals.

I spent several hours to try to work around the test service and I'd probably need more time to review whether we can remove the test service.

Therefore, the easiest solution that I see is to leave LiteralText as public so we come across issues with the Microsoft.OData.Service test project; in trying to solve this issue, I noticed that a few parts of the production code has `#if ODATA_SERVICE` defines just to cater to the test project which I don't think is a good thing. Perhaps the FilterSegment interface could be re-evaluated during the next major release.